### PR TITLE
feature/lisasiliu/add-programs-form-backend-route

### LINKED ISSUE #20: EPIC - Add Programs Form
An important functionality of the app is to be able to add programs to the database. Programs represent the classes that students sign up for. For this feature we need a form to input data and our backend needs to update the data in our database. The figma designs can be found [here.](https://www.figma.com/file/N16keN1mV1QDnorLB1GC6C/Plant-it-Again-(PIA)-Program-Manager?type=design&node-id=2020-9410&mode=design&t=WnlTGgLHoaEgPgqq-0)

Tasks:

- [x] [Create a backend route to add all necessary information about a program](https://github.com/TritonSE/PIA-Program-Manager/issues/9)
- [x] [Add a UI form for creating a new program](https://github.com/TritonSE/PIA-Program-Manager/issues/30)
- [x] Connect Frontend and Backend, ensure that data inputted on frontend is stored on backend

### PR DESCRIPTION
**Pull request by lisasiliu**
## Tracking Info

Resolves #20 

## Changes

Added program form schema, validator, createProgramForm, post route

## Testing

Tested incorrect fields, missing fields

## Confirmation of Change

See files in backend/src. Request bodies use following structure:
{
    "name": "A New Program",
    "abbreviation": "ANP",
    "type": "regular",
    "daysOfWeek": ["M", "W", "TH"],
    "startDate": "2022-09-10T11:00:55Z",
    "endDate": "2023-08-10T11:00:55Z",
    "color": "#FFFFFF"
}

### CODE REVIEW

**Comment by lisasiliu** [comment]
there are still type cast errors to be fixed

**Comment by lisasiliu** [comment]
ready for review!! :DD

**Review by adhi0331 — APPROVED** [review]
LGTM!!! 🚀

**Comment by JacobLee23** [comment]
Ran some positive and negative test combinations for the `/program` route.

In general, the route seems to pass and fail appropriately, with the exception of the `color` field:

image

The bug seems to originate from the condition in line 71:

```ts
const makeColorValidator = () => 
  body("color")
    ...
    .custom((value: string) => {
      ...
      if (value.startsWith("#")) throw new Error("color hex should start with #");
    });
```

If I'm not mistaken, the condition should be `!value.startsWith("#")`.

Additionally, an additional validator should be implemented to throw an error when the `type` field is not `"regular"` or `"varying"`:

image

In the other cases I tried, the route seemed to function as expected:

image
image

File: backend/src/validators/program.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    .withMessage("color hex should be string")
+    .isLength({ min: 7, max: 7 })
+    .withMessage("color hex should have 7 characters")
+    .bail()
+    .notEmpty()
+    .withMessage("color hex required")
+    .bail()
+    .custom((value: string) => {
+      if (value.length !== 7) throw new Error("color hex should have 7 characters");
+      if (value.startsWith("#")) throw new Error("color hex should start with #");
```
Comment (JacobLee23) [inline backend/src/validators/program.ts]: Possibly:

```ts
if (!value.startsWith("#")) ...
```

File: backend/src/validators/program.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+const makeTypeValidator = () =>
+  body("type")
+    .exists()
+    .withMessage("type needed")
+    .bail()
+    .isString()
+    .withMessage("type must be a string")
+    .bail()
+    .notEmpty()
+    .withMessage("type must not be empty");
```
Comment (JacobLee23) [inline backend/src/validators/program.ts]: Custom validator to restrict values of `type` field:

```ts
.custom((value: string) => { value === "regular" | | value === "varying" });
```

**Comment by lisasiliu** [comment]
bugs mentioned are fixed @JacobLee23

**Comment by JacobLee23** [comment]
Changes look good to me!

image
image
image

**Review by JacobLee23 — APPROVED** [review]

**Pull request merged**